sympy/tensor/tensor.py

```diff
@@ -921,6 +925,14 @@ def set_comm(self, i, j, c):
         self._comm[ni][nj] = c
         self._comm[nj][ni] = c
 
+        """
+        Cached sympy functions (e.g. expand) may have cached the results of
+        expressions involving tensors, but those results may not be valid after
+        changing the commutation properties. To stay on the safe side, we clear
+        the cache of all functions.
+        """
+        clear_cache()
```

Comment:
> Maybe the commutation properties should be made unmodifiable after the tensor object has been created?

I am not completely sure what the best way to implement it would be, though.

The obvious solution would be to modify `_TensorManager.set_comm` so that it raises an error if the user uses it to modify the commutation properties of pre-existing commutation groups, but I can see how that would make more complex use-cases impossible (even the current docstring of `_TensorManager.set_comm` has an example where it is necessary to modify the commutation properties of pre-existing groups).

Another possibility would be to maintain a list (say an attribute `_frozen_comms`) in the `_TensorManager` class, such that every TensorHead will add its commutation group to this list when it is instantiated. `_TensorManager.set_comm` would then raise an error when the user tries to modify the properties of any commutation group that is in this list. Do you see any problems in this approach? If not, I can try to implement it any then make another pull request.

Reply:
> Another possibility would be to maintain a list (say an attribute _frozen_comms) in the _TensorManager class, such that every TensorHead will add its commutation group to this list when it is instantiated. _TensorManager.set_comm would then raise an error when the user tries to modify the properties of any commutation group that is in this list. Do you see any problems in this approach? If not, I can try to implement it.

I can imagine this would make interactive use a bit painful, though. For example, let's say I have done some calculation with a bunch of TensorHead instances that belong to a particular commutation group 'A', and suddenly I decide that I want to define a new TensorHead in a group 'B' that anticommutes with those in group 'A'. This would not be possible if I were to implement the `_frozen_comms` attribute, and I would be forced to open a new Sympy session, define the new commutation group, and then rerun all my old calculations.

Reply:
> I can imagine this would make interactive use a bit painful, though....

I'll try to come up with an implementation over the weekend and check if this is really the case.

Reply:
@Kishore96in if you want, we can still merge this PR as a temporary fix to the issue. Let me know.

Reply:
> @Kishore96in if you want, we can still merge this PR as a temporary fix to the issue. Let me know.

@Upabjojr yes, I would like that.

A question about implementing immutability for the TensorHead commutation properties: do we need to deprecate modification of the commutation properties after instantiation, or is it okay to directly remove the ability to modify it?

Reply:
> do we need to first deprecate modification of the commutation properties after instantiation, or is it okay to directly remove the ability to modify it?

We need deprecation first. Many users use SymPy, and they should be warned about coming API changes.

Reply:
Then I think the cache clearing hack here will anyway be needed during the deprecation period. I would suggest merging this now.

I'll make another pull request that emits a deprecation warning when the user tries to change commutation properties of a pre-existing tensor (this might take a while, since I will have to think about the simplest way to implement it).

Reply:
Merged. Thank you for your contribution!